flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitMonitoringFunction.java

```diff
@@ -0,0 +1,345 @@
+				monitorDirAndForwardSplits(fileSystem, context);
+				isRunning = false;
+				break;
+			default:
+				isRunning = false;
+				throw new RuntimeException("Unknown WatchType" + watchType);
+		}
+	}
+
+	public boolean isRunning() {
+		return this.isRunning;
+	}
+
+	private void monitorDirAndForwardSplits(FileSystem fs, SourceContext<FileInputSplit> context) throws IOException, JobException {
+		final Object lock = context.getCheckpointLock();
+
+		// it may be non-null in the case of a recovery after a failure.
+		if (currentSplitsToFwd != null) {
+			synchronized (lock) {
+				forwardSplits(currentSplitsToFwd, context);
+			}
+		}
+		currentSplitsToFwd = null;
+
+		// it may be non-null in the case of a recovery after a failure.
+		if (splitsToFwdOrderedAscByModTime == null) {
+			splitsToFwdOrderedAscByModTime = getInputSplitSortedOnModTime(fs);
+		}
+
+		Iterator<Tuple2<Long, List<FileInputSplit>>> it =
+			splitsToFwdOrderedAscByModTime.iterator();
+
+		while (it.hasNext()) {
+			synchronized (lock) {
+				currentSplitsToFwd = it.next();
+				it.remove();
+				forwardSplits(currentSplitsToFwd, context);
+			}
+		}
+
+		// set them to null to distinguish from a restore.
+		splitsToFwdOrderedAscByModTime = null;
+		currentSplitsToFwd = null;
+	}
+
+	private void forwardSplits(Tuple2<Long, List<FileInputSplit>> splitsToFwd, SourceContext<FileInputSplit> context) {
+		currentSplitsToFwd = splitsToFwd;
+		Long modTime = currentSplitsToFwd.f0;
+		List<FileInputSplit> splits = currentSplitsToFwd.f1;
+
+		Iterator<FileInputSplit> it = splits.iterator();
+		while (it.hasNext()) {
+			FileInputSplit split = it.next();
+			processSplit(split, context);
+			it.remove();
+		}
+
+		// update the global modification time
+		if (modTime >= globalModificationTime) {
+			globalModificationTime = modTime;
+		}
+	}
+
+	private void processSplit(FileInputSplit split, SourceContext<FileInputSplit> context) {
+		LOG.info("Forwarding split: " + split);
+		context.collect(split);
+	}
+
+	private List<Tuple2<Long, List<FileInputSplit>>> getInputSplitSortedOnModTime(FileSystem fileSystem) throws IOException {
+		List<FileStatus> eligibleFiles = listEligibleFiles(fileSystem);
+		if (eligibleFiles.isEmpty()) {
+			return new ArrayList<>();
+		}
+
+		Map<Long, List<FileInputSplit>> splitsToForward = getInputSplits(eligibleFiles);
+		List<Tuple2<Long, List<FileInputSplit>>> sortedSplitsToForward = new ArrayList<>();
+
+		for (Map.Entry<Long, List<FileInputSplit>> entry : splitsToForward.entrySet()) {
+			sortedSplitsToForward.add(new Tuple2<>(entry.getKey(), entry.getValue()));
+		}
+
+		Collections.sort(sortedSplitsToForward, new Comparator<Tuple2<Long, List<FileInputSplit>>>() {
+			@Override
+			public int compare(Tuple2<Long, List<FileInputSplit>> o1, Tuple2<Long, List<FileInputSplit>> o2) {
+				return (int) (o1.f0 - o2.f0);
+			}
+		});
+
+		return sortedSplitsToForward;
+	}
+
+	/**
+	 * Creates the input splits for the path to be assigned to the downstream tasks.
```

Comment:
The `files` parameter should be called `eligibleFiles`, right? Also the comment seems to indicate that files in the list are not used while they are the ones that are actually used.